### DISCUSSION TITLE
[general] Scratch file growing out of control (inconsistently)

### DISCUSSION THREAD

**Discussion by sakrejda-umich** [ropensci/targets#1204] [General]
### Help

- [X] I understand and agree to <https://books.ropensci.org/targets/help.html>.

### Description

I'd appreciate a high-level pointer troubleshooting a large pipeline. I have targets that run fine in a clean session (with the deps loaded using `targets::tar_load` as needed) but when I run through `tar_make` a few will, inconsistently, run, apparently finish (or at least the output looks like it all gets produced) and then start writing a file in the `scratch` directory... and then just keep writing it... and eat the entire hard drive. These are not very large targets... so the pointer I'm looking for: if I were to look for where in the `targets` code the writing of scratch files gets triggered towards the end of processing a target... where do I look for that?

I did create one of these files with permissions that the main user was not able to write to and then used `options(error=recover)` to look for whatever function is doing the writing but haven't had luck finding that yet. I will file a more complete issue when I understand the problem a little more but I'd appreciate any pointers to the code that utilizes scratch files

**Comment by wlandau** [comment]
> when I run through tar_make a few will, inconsistently, run, apparently finish (or at least the output looks like it all gets produced) and then start writing a file in the scratch directory... and then just keep writing it... and eat the entire hard drive. These are not very large targets... 

Do you mean the same file gets written repeatedly to `_targets/scratch/`, or that multiple files for different targets accumulate there? Or, do you mean a single target in a single `tar_make()` writes multiple times to `_targets/scratch/`, either to the same or a different path?

> so the pointer I'm looking for: if I were to look for where in the targets code the writing of scratch files gets triggered towards the end of processing a target... where do I look for that?

The most helpful thing would be a reproducible example so I can run the code and understand the clarifying details of the problem. 

The method that writes to `_targets/scratch/` is:

https://github.com/ropensci/targets/blob/851cfc9105ec92a972de912fcc88ce03fa940c1d/R/class_store.R#L127-L134

However, the code code base of `targets` is highly object-oriented, and there are different methods for `store_write_path()` and `store_convert_object()` for different storage formats (see the "format" argument of `tar_target()`). Example for `format = "rds"`:

https://github.com/ropensci/targets/blob/851cfc9105ec92a972de912fcc88ce03fa940c1d/R/class_rds.R#L15-L18

**Comment by sakrejda-umich** [comment]
Thanks for the hints! Just for clarity: the target creates a single file,
the file grows until the disk runs out of space. I have a good guess about
what's up. Let me check a few more possibilities and I should be able to
either tell if the issue is in targets code or one of the other packages it
relies on, then I'll file an issue.

**Comment by sakrejda-umich** [comment]
Ok, I think this can be closed as not-`targets`: for *some* ggplot objects `qs::qsave` will write arbitrarily large files. My current guess is it's finding the data's environment and if that environment also contains the plot it just keeps going, but I can reproduce without `targets`
